File: linksortadder.py
class Node:
    def __init__(self, body, next):
        if body:
            self.body = body
        else:
            self.body = 0
        if next:
            self.next = next
        else:
            self.next = None


def main():
    n2 = Node(8, 
         Node(8, 
         Node(19, 
    None)))

    n1 = Node(-4, 
         Node(2, 
         Node(8, 
    None)))

    printlist(n1)
    printlist(n2)
    start = synchronize(n1, n2)
    printlist(start)


def printlist(start):
    while start != None:
        print(start.body, end=" ")
        start = start.next
    print()


def synchronize(node1, node2):
    start = None
    s = None
    l1 = None
    l2 = None

    if node2.body > node1.body:
        start = node1
        s0 = node1
        s1 = node1.next
        l1 = node1.next
        l2 = node2
    else:
        start = node2
        s0 = node2
        s1 = node2.next
        l1 = node1
        l2 = node2.next

    while(l1 != None and l2 != None):
        if (l1.body > l2.body):
            s1 = l2
            s0.next = s1
            s0 = s0.next
            l2 = l2.next
        else:
            s1 = l1
            s0.next = s1
            s0 = s0.next
            l1 = l1.next
    
    while (l1 != None):
        s1 = l1
        s0.next = s1
        s0 = s0.next
        l1 = l1.next
    while (l2 != None):
        s1 = l2
        s0.next = s1
        s0 = s0.next
        l2 = l2.next

    return start


if __name__ == "__main__":
    main()

File: linksortadder.cpp
#include <iostream>
using namespace std;

typedef struct {
    int body;
    Node next;
} Node;


int main() {

    Node n6; n6.body = 6; n6.next = nullptr;
    Node n4; n4.body = 4; n4.next = &n6;
    Node n2; n2.body = 2; n2.next = &n4;

    Node n5; n5.body = 5; n5.next = nullptr;
    Node n3; n3.body = 3; n3.next = &n5;
    Node n1; n1.body = 1; n1.next = &n3;

    Node* start = synchronize(n1, n2);
    printlist(start);

}


void printlist(Node* start) {

    while (start != nullptr) {
        cout << start -> body;
        start = start -> next;
    }

}

Node* synchronize(Node node1, Node node2) {

    Node* start = nullptr;
    Node* s = start;
    Node* l1 = &node1;
    Node* l2 = &node2;

    while (true) {
        if (l1 == nullptr) {
            if (l2 == nullptr) return start;
            while (l2 != nullptr) {
                s -> next = l2;
                s = s -> next;
                cout << s -> body;
                l2 = l2 -> next;
            }
            return start;
        }
        else if (l2 == nullptr) {
            while (l1 != nullptr) {
                s -> next = l1;
                s = s -> next;
                cout << s -> body;
                l1 = l1 -> next;
            }
            return start;
        }

        if (l1 -> body > l2 -> body) {
            s -> next = l1;
            s = s -> next;
            l1 = l1 -> next;
        }
        else if (l1 -> body < l2 -> body) {
            s -> next = l2;
            s = s -> next;
            l2 = l2 -> next;
        }
    }

}